### DISCUSSION TITLE
Pocketbase Framework (jsvm): $http.send and setting cookies.

### DISCUSSION THREAD

**Discussion by jonshipman** [pocketbase/pocketbase#4178] [Q&A] [answered]
Porting some existing application over to the pocketbase framework by way of hooks and noticing that `$http.send({headers:{Cookie: "cookie string"}});` is not taking. I can read cookies by either `response.cookies` or `response.headers['Set-Cookie']` but for some reason the server I'm sending to isn't acknowledging the cookie.

I know this is a shot in the dark as there are too many variables at play, but firing up the old application in nodejs was able to pass down headers.Cookie in the https.request method. Just wondering if there's anything internal with the golang implementation that's well known that I'm doing incorrectly.

**Comment by ganigeorgiev** [comment]
I'm not sure what could be the cause of the issue. It works fine for me when testing locally with:

```js
routerAdd("GET", "/test", (c) => {
    const cookie = c.cookie("test")

    return c.json(200, { cookie })
})

routerAdd("GET", "/send", (c) => {
    const res = $http.send({
        url:     "http://127.0.0.1:8090/test",
        method:  "GET",
        headers: {
            "content-type": "application/json",
            "cookie": "test=example"
        },
    })

    return c.json(200, { result: res.json.cookie })
})
```

Navigate to `http://127.0.0.1:8090/send` and you should see the parsed cookie from the `/test` route.

It is possible that your server may have some requirements for the cookie key and value and/or expect some specific cookie attribute like `maxAge`, `domain`, `httpOnly`, `secure`, etc. (_in the above example all of the attributes are skipped and you should see their zero-defaults_).

**Reply by jonshipman** [reply]
Is `application/x-www-form-urlencoded` supported? That's the only thing I can think of that's different as URLSearchParams doesn't work in the jsvm. I am doing `${encodeURIComponent(key)}=${encodeURIComponent(value)}` which should theoretically work the same

**Reply by ganigeorgiev** [reply]
I'm confused. I don't think the request content type, nor `URLSearchParams` has anything to do with cookies. Could you provide a _minimal_ code sample showing how you are sending the cookie and, if PocketBase is also the receiving end, the route that reads it?

**Reply by jonshipman** [reply]
I was able to reproduce the issue in node's fetch. If I recall correctly, it has to do with redirects (node's fetch will automatically handle redirects while `https.request` will not). Is there an option for `$https.send` to ignore redirects?

This example isn't in pocketbase, but it illustrates the difference
```
import https from 'node:https';

await fetch('https://my.acculynx.com/Signin/Signin').then((r) => {
  console.log(r.status);
});

https
  .request(
    {
      host: 'my.acculynx.com',
      path: '/Signin/Signin'
    },
    (r) => {
      console.log(r.statusCode);
    }
  )
  .end();

```

You'll get a `200` from the fetch and a `301` from https.request.

**Reply by ganigeorgiev** [reply, marked as answer]
Indeed redirects could be the cause for the lack of cookies. The std Go http client forwards most headers when following redirects but there are [some requirements](https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/net/http/client.go;l=44) for the sensitive ones:

> // When forwarding sensitive headers like "Authorization", "WWW-Authenticate", and "Cookie" to untrusted targets.
> // These headers will be ignored when following a redirect to a domain
> // that is not a subdomain match or exact match of the initial domain.
> // For example, a redirect from "foo.com" to either "foo.com" or "sub.foo.com"
> // will forward the sensitive headers, but a redirect to "bar.com" will not.

---

>  Is there an option for $https.send to ignore redirects?

No, at the moment there is no such option, but even if there was I'm not sure how this will work in your use case. Are you planning to recursively and manually send requests with the cookie to each redirect target?

**Reply by jonshipman** [reply]
Correct. Client needs some data analytics from their CRM that aren't accessible via their (acculynx) API. The process is as follows:

1. Initial request to establish session cookie.
2. POST request to their signin form which returns a .ASPXAUTH session token.
3. This token can be used to fetch any JSON inside the application.

Good to know it's not anything inherently broken within their system that enables this, only that the redirects are skipping the cookies. This application would technically be "untusted" as it's running on a server and not via a web browser. We have been using a system in Google BigQuery for over a year but transitioning a lot of their applications over to Pocketbase because of the added freedom.

To solve this issue, I'm just running a round robin in a separate sveltekit application that does the old https.request logic and passes it back to Pocketbase. Setup as a class where each request stores the response's cookies and those cookies are used in subsequent requests. Login isn't persistent across the application, just for the current request I'm performing. Works well as this is a simple caching application that allows other applications access to this data. The fetches are run on a cronAdd function periodically.
